monitoring/tests/test_publishers.py

```diff
@@ -0,0 +1,110 @@
+        self.assertEqual(connection_str, 'mysql://admin:test123@1.2.3.4/TestDB')
+
+
+class TestPublishing(unittest.TestCase):
+    def setUp(self):
+        # Provision the SQLite in-memory database with our schema
+        self.engine = create_engine('sqlite://')
+        MonitoringBase.metadata.create_all(self.engine)
+
+        self.session_factory = sessionmaker()
+        self.session_factory.configure(bind=self.engine)
+        self.session = self.session_factory()
+
+    def tearDown(self):
+        self.session = None
+        self.session_factory = None
+        self.engine.dispose()
+        self.engine = None
+
+    def test_publish_build_stats_no_stages(self):
+        build_stats = BuildStats(
+            job_name = 'Test',
+            build_id = 1234,
+            build_url = 'http://localhost:8080/job/MyJob/1',
+            finished_at_secs = 5678,
+            status = 'SUCCESS',
+            duration_millis = 4321,
+            stages = [],
+        )
+
+        with patch.object(MySQLPublisherAdapter, 'get_database_session', return_value=self.session):
+            publisher = MySQLPublisherAdapter('', '', '', '')
+            publisher.publish(build_stats)
+
+        build_count = self.session.query(Build).count()
+        self.assertEqual(build_count, 1)
+
+        stage_count = self.session.query(Stage).count()
+        self.assertEqual(stage_count, 0)
+
+        build = self.session.query(Build).one()
+        self.assertEqual(build.job_name, 'Test')
+        self.assertEqual(build.build_id, 1234)
+        self.assertEqual(build.build_url, 'http://localhost:8080/job/MyJob/1')
+        self.assertEqual(build.finished_at_secs, 5678)
+        self.assertEqual(build.status, 'SUCCESS')
+        self.assertEqual(build.duration_millis, 4321)
+
+    def test_publish_build_stats_with_stages(self):
+        stage1_stats = StageStats(
+            name = 'TestStage1',
+            status = 'SUCCESS',
+            duration_millis = 1010,
+        )
+
+        stage2_stats = StageStats(
+            name = 'TestStage2',
+            status = 'FAILURE',
+            duration_millis = 10,
+        )
+
+        build_stats = BuildStats(
+            job_name = 'Test',
+            build_id = 1234,
+            build_url = 'http://localhost:8080/job/MyJob/1',
+            finished_at_secs = 5678,
+            status = 'SUCCESS',
+            duration_millis = 4321,
+            stages = [stage1_stats, stage2_stats],
+        )
+
+        with patch.object(MySQLPublisherAdapter, 'get_database_session', return_value=self.session):
+            publisher = MySQLPublisherAdapter('', '', '', '')
+            publisher.publish(build_stats)
+
+        build_count = self.session.query(Build).count()
+        self.assertEqual(build_count, 1)
+        build = self.session.query(Build).one()
+        self.assertEqual(len(build.stages), 2)
+        self.assertIsNotNone(build.stages[0])
+        self.assertIsNotNone(build.stages[1])
```

Comment:
Unit tests should ideally be like 5 lines long.

Maybe we should do some helper functions, like `assert_stages_equal(stage1_stats, build.stages[0])`.

Also, maybe we should move test objects, like `stage2_stats` into set_up or just as a class field.

The goal here is to make the actual test body as short and clear as possible.

This goes for all other such "long" tests in this PR. 

Then, this test would look like:
```
build_stats_with_two_stages = (..., stages=[...])

...

def test_publish_build_stats_with_stages(self):
    mock_something_according_to_other_PR_comment()
    publisher = MySQLPublisherAdapter('', '', '', '')
    publisher.publish(build_stats_with_two_stages)
    assert_has_two_stages(build)
    assert_stages_equal(stage1_stats, build.stages[0])
    assert_stages_equal(stage2_stats, build.stages[1])
```